Added symbol table

jenrik opened the pull request:
Added AST visitor, and base implementation of the visitor, see 
a71d6f0.
Added a AST walker that extract values from the parse tree, see 0059a2e.
Added a symbol table.
Added symbol walker that extract symbols and does declaration-before-use checking, 
d75818d.

jenrik committed:
Resolved ToDo in mapper.kt

The child node of Number_literalContext is either integer or float

jenrik committed:
Added AST visitor

Added a visitor interface for the AST, as well as a default
implementation

jenrik committed:
Added value extraction walker

Added a tree walker that extract values like identifiers, literals, and
neighbourhood coordinates

jenrik committed:
Added symbol walker

Added a tree walker that extracts symbols, and checks for use of symbols
before they are declared

jenrik committed:
Added reserved symbols

Language keywords and builtin functions can no longer be used as
variable, state, constant, or function names

jenrik commented:
The symbol table now rejects attempts to insert symbols that matches language keywords, and builtin functions

NicEastvillage commented:
`.circleci/config.yml` does not seem relevant for this change.

NicEastvillage commented on compiler/src/main/kotlin/ast/symbol.kt:
```diff
@@ -0,0 +1,82 @@
+    "floor",
+    "ceil",
+    "sqrt",
+    "pow"
```
pow, sqrt, ceil og andre indbyggede funktioner er ikke reserverede nøgleord. Disse funktioner skal lægges i symbol table før brug. Der skal muligvis et eller andet hacky til for at lave en AST node til dem. Umiddelbart tænker jeg vi kan lave en unik AST node til hver af dem.

jenrik replied:
I think we should take care of this in a future pull request since resolving it is a rather complex matter involving how we should handle function overloading.

falkecarlsen replied:
In that case, unless @NicEastvillage disagrees, we should make an issue to track the work.

jenrik commented:
the test and CircleCI related commits has been removed

NicEastvillage commented on compiler/src/main/kotlin/walkers/value.kt:
```diff
@@ -0,0 +1,133 @@
+class CoordinateParsingException: ParsingException("Inconsistent amount axes in neighbourhood")
+
+
+class ParseTreeValueWalker : BaseASTVisitor() {
```
Jeg ville kalde denne for LiteralExtractorVisitor

NicEastvillage commented on compiler/src/main/kotlin/walkers/symbol.kt:
```diff
@@ -0,0 +1,60 @@
+
+class SymbolException(val ident: String) : Exception("\"$ident\" was used before it was declared")
+
+class SymbolWalker(val symbolTable: SymbolTable) : BaseASTVisitor() {
```
Dette ligner ScopeCheckVisitor.

NicEastvillage reviewed (dismissed):
Jeg har nogle ændringer til hvordan indbyggede funktioner gemmes i symbol table (dette kan vi ændre i et nyt PR), og nogle forslag til navneændringer.  Men ellers ser indholdet korrekt ud.

NicEastvillage commented:
Btw, make sure to negative numbers are extracted as negative values in LiteralExtractorVisitor (aka ParseTreeValueWalker). PR #41 will change the definition of integer_literals and float_literals.

jenrik commented:
The conversion from number literals to there respective value is done through kotlin's `String.toInt()` and `String.toFloat()` that includes negative number. So if the integer and float tokens consume the minus kotlin will interpret them as negative.

jenrik committed:
Renamed ParseTreeValueWalker and SymbolWalker

Renamed ParseTreeValueWalker to LiteralExtractorValue.
Renamed SymbolWalker to ScopeCheckVisistor.

falkecarlsen reviewed (dismissed):
Looks good, all relevant comments have been resolved.

Please write an issue regarding the built-in functions symbols, though.

NicEastvillage commented on compiler/src/main/kotlin/walkers/symbol.kt:
```diff
@@ -0,0 +1,60 @@
+        super.visit(node)
+        symbolTable.closeScope()
+    }
+}
```
You have VarExpr where symbol declaration is checked, but FuncExpr, StateExpr, NeighbourhoodExpr (or whatever their names are) are not checked if the name is already declared.

jenrik replied:
State names and neighbourhood names are currently matched by VarExpr, so maybe we should rename VarExpr to something more generic.

jenrik replied:
e662dc1 implements code that checks functions are declared before they are used.

NicEastvillage replied:
I suggest `Identifier` for named expressions. Or `NamedExpression`.

NicEastvillage replied:
I will resolve and approve after rename.

jenrik replied:
I have renamed `VarExpr` to `NamedExpr` in 8247d7f

falkecarlsen commented on compiler/src/main/kotlin/walkers/symbol.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    override fun visit(node: StateDecl) {
+        symbolTable.createScope()
```
Does not check if the color declaration contain undeclared constants.

jenrik replied:
I think we should discus what expressions are allowed in state colors.

jenrik replied:
I looked into the grammar and found that colors currently are limited to integer literals, so we should fill issue regarding expression as colors and the limitations we probably gonna have to put on them.

jenrik committed:
Check for declaration-before-use of functions

falkecarlsen reviewed (changes requested):
Another missing functionality, can be put off, if necessary.

jenrik committed:
Renamed VarExpr to NamedExpr

Renamed VarExpr to NamedExpr since it covers both variables, states,
constants, and neighbourhoods.
